### PR TITLE
Fix EntitiesSavedStates panel dialog props.

### LINKED ISSUE #67313: Entity saved modal dialog closes unexpectedly
### Description

When clicking the 'padding area' of the Entity saved modal dialog, the modal dialog closes unexpectedly.

Animated GIF to illustrate:

Image

Turns out this modal dialog content, the `EntitiesSavedStatesExtensible` component, uses the `useDialog` hook to attach modal behaviors to _the content_. It is then used inside a Modal dialog. Basically, from a functional perspective, this is a Modal within a Modal.

That explains why clicking outside the 'inner modal' (the content) closes the modal. In fact, clicking outside the area highlighted in the screenshot below but still inside the 'outer modal' closes the modal.

Image

### Step-by-step reproduction instructions

- Go to the Site editor.
- Make a simple change to a global style, e.g. change a color.
- Reopen the Navigation panel by clicking 'Open Navigation' at the top left of the screen.
- Click any item in the navigation other than 'Styles' e.g. click 'Pages' or 'Templates'.
- At this point, a blue button at the bottom of the Navigation panel appears, labeled 'Review 1 change...'

Image

- Click the button: the modal dialog to save entities opens.
- Click the padding area of the modal dialog.
- Observe the modal dialog closes.
- Alternatively preview a block theme from the WP admin > Appearance > any block theme > Live Preview.
- Click the blue button labeled 'Activate {theme name}' at the bottom of hte Navigation panel: the modal dialog to save entities opens.
- Click the padding area of the modal dialog.
- Observe the modal dialog closes.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [ ] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

### PR DESCRIPTION
**Pull request by afercia**
Fixes https://github.com/WordPress/gutenberg/issues/67313

Alternative to https://github.com/WordPress/gutenberg/pull/67327

## What?

When the entity saved states component is rendered inside a Modal component, clicking the padding area inside the modal dialog closes the modal unexpectedly.

Basically, this PR seeks to:
- Pass down the `renderDialog` prop as it is supposed to be used.
- Attach the `useDialog` props to the component only when it is expected to work with a modal behavior i.e. when `renderDialog` is true.
- Improve the `renderDialog` docs.

## TL;DR

https://github.com/WordPress/gutenberg/pull/59622 made the `EntitiesSavedStates` component behave like a modal dialog when the new `renderDialog` prop is true. However, the props from `useDialog` [are always passed unconditionally](https://github.com/WordPress/gutenberg/blob/443ce894391a0e07738661c5b3d4af4e34c47616/packages/editor/src/components/entities-saved-states/index.js#L123-L124) to the component. As such, the component gets some modal behaviors including the 'close when clicking outside'. This works when the component is rendered in the save panel on the right. I doesn't work well when it is rendered inside a Modal dialog because in this case it's a sort of 'modal behavior' within a 'modal behavior' e.g.: clicking outside the inner component with modal behavior but still inside the outer Modal component closes the modal unexpectedly.

## Why?

Modal dialogs should not close unexpectedly.
The `useDialog` props should be passed conditionally, in the same way `role=dialog`, `aria-labelledby`, and `aria-describedby` are passed conditionally. This makes sure that when the component gets the ARIA attributes related to a modal dialog, it also behaves like a modal dialog. When `renderDialog` is false, none of the props related to the modal dialog should be set.

## How?

- Passes the `useDialog` props conditionally based on `renderDialog`.
- Fixes the `renderDialog` props, please double check this as it seems to me the docs introduced in https://github.com/WordPress/gutenberg/pull/62377 aren't accurate.
- Makes sure that the buttons that open the entity saved state panel do have an `aria-haspopup="dialog"` attribute.

## Testing Instructions

- Repeat the testing instructions from https://github.com/WordPress/gutenberg/issues/67313
- Observe the modal dialog does not close unexpectedly any longer.
- Make styles changes in the Site editor or preview a block theme in the editor.
- Click the 'Save' or 'activate theme' button at the top right of the screen: the entity saved state component renders in the right sidebar.
- Inspect the source and observe the element with class `entities-saved-states__panel` does have a role=dialog and correct aria-labelledby and aria-describedby.
- Note: when there are changes, this component may still be rendered in the DOM even when it is closed and it's hidden via CSS.

These are the ARIA properties for the component in the Site editor:

site editor

And these are the ones for the theme preview:

theme preview

- Check that wherever the entity saved state component renders in the right sidebar:
  - It closes when clicking outside of it.
  - It has correct aria role, aria-labelledby and aria-describedby.
- For example, this should be checked also in the Post editor, after editing a template from the post editor and saving, the entity saved state component is used instead of the normal publish panel.

### CODE REVIEW

**Comment by afercia** [comment]
Cc @WordPress/gutenberg-core I'd appreciate a review, when you have a chance.

File: packages/editor/src/components/entities-saved-states/index.js [hunk 1]
```diff
@@ -72,7 +69,7 @@ export function EntitiesSavedStatesExtensible( {
 	onSave = identity,
 	saveEnabled: saveEnabledProp = undefined,
 	saveLabel = __( 'Save' ),
-	renderDialog = undefined,
+	renderDialog = false,
```
Comment (ntsekouras) [inline packages/editor/src/components/entities-saved-states/index.js]: Why change to `false`? Did it affect something by being `undefined`? I mean we could even remove the ` = undefined` part, no?

Comment (afercia) [reply packages/editor/src/components/entities-saved-states/index.js]: Yes I considered to remove the ` = undefined`. I thought it's more explicit this way as it's a 'flag'. No strong opinions though.

Comment (ntsekouras) [reply packages/editor/src/components/entities-saved-states/index.js]: I think it's fine to remove.

**Comment by afercia** [comment]
Removed `renderDialog` unnecessary default value and rebased.

**Review by ntsekouras — APPROVED** [review]
LGTM, thanks!

File: packages/editor/README.md [hunk 2]
```diff
@@ -401,7 +401,7 @@ _Parameters_
 
 -   _props_ `Object`: The component props.
 -   _props.close_ `Function`: The function to close the dialog.
--   _props.renderDialog_ `Function`: The function to render the dialog.
+-   _props.renderDialog_ `boolean`: Whether to render the component with modal dialog behavior.
```
Comment (ciampo) [inline packages/editor/README.md:404]: Would this qualify as a breaking change?

(sorry for the delayed comment, I'm still catching up with notifications from the past weeks)

Comment (ntsekouras) [reply packages/editor/README.md:404]: No, the documentation was wrong when added. The value was always boolean.

Comment (Mamaduka) [reply packages/editor/README.md:404]: IMO, this component shouldn't be public, but that ship has sailed :(

**Comment by Mamaduka** [comment]
The 'close when clicking outside' behavior no longer works in the post editor, but it does in the site editor. I'm not sure if it's intentional or a recent regression.

**Comment by afercia** [comment]
> The 'close when clicking outside' behavior no longer works in the post editor

It works for me. Are you testing the entity saved panel or the normal pubish panel? As in: in the post editor you need to edit something other than the post content to get the enity saved panel e.g. edit the template from the post editor.

**Comment by afercia** [comment]
GIF to illustrate the entity saved panel in the post editor 'close when clicking outside' behavior:

entity

**Comment by Mamaduka** [comment]
Sorry, I was testing the normal publish panel. However, I expect it to have similar behavior since they share the design.

**Comment by afercia** [comment]
> I expect it to have similar behavior since they share the design.

I'd agree and I noted it in the issue:

> However, this is inconsistent with other similar panels. For example, the pre-publish checks panel doesn't close when clicking outside of it.

Also, see [this comment](https://github.com/WordPress/gutenberg/issues/67313#issuecomment-2505518921) where there's some related considerations.
